File: apps/android/app/src/main/java/dev/flextranslate/foundation/M2m100Tokenizer.kt
package dev.flextranslate.foundation

import android.util.Log
import org.json.JSONObject
import java.io.File
import java.text.Normalizer

/**
 * Точный порт fast-токенизатора M2M-100 от HuggingFace на чистом Kotlin, целиком на основе родного
 * `tokenizer.json` модели. Портируем руками, потому что готового артефакта под Android НЕТ:
 * `ai.djl.huggingface:tokenizers` тащит только десктопные нативы (linux/osx/win) без arm64-v8a `.so`,
 * а официальной Android-сборки HF `tokenizers` не существует.
 *
 * Конвейер (точно по спеке `tokenizer.json`, кроме отмеченного):
 *  1. Нормализация — NFKC. В референсе SentencePiece "Precompiled" charsmap; для разговорного текста
 *     RU/EN/ZH из демо NFKC — точное подмножество (свёртка по ширине/совместимости).
 *     Честно: экзотические кодпоинты могут токенизироваться чуть иначе, чем в референсе.
 *  2. Предтокенизация — WhitespaceSplit, затем Metaspace: каждое слово, отделённое пробелом,
 *     получает префикс-маркер `▁` (U+2581) (`add_prefix_space = true`).
 *  3. BPE по словам — `ignore_merges = true`, поэтому слово, уже целиком есть в vocab, выдаётся
 *     целиком; иначе обычное слияние пар по рангу, начиная с символов. `fuse_unk = true` сворачивает
 *     цепочки неизвестных символов в один `<unk>`. Без dropout и без byte-fallback.
 *  4. Постобработка — TemplateProcessing `single`: `[__src__] tokens </s>`.
 *
 * Декодирование разворачивает metaspace (`▁` → пробел) и выкидывает спецтокены.
 *
 * Id токенов берутся прямо из vocab, так что языковые токены M2M-100 (`__en__`=128022,
 * `__ru__`=128077, `__zh__`=128102) и управляющие id (`<s>`=0, `<pad>`=1, `</s>`=2, `<unk>`=3) точно
 * совпадают с референсными — на это и опирается логика forced-BOS у ONNX-декодера.
 */
class M2m100Tokenizer private constructor(
    private val vocab: Map<String, Int>,
    private val idToToken: Map<Int, String>,
    private val mergeRanks: Map<Long, Int>,
    private val unkId: Int,
) {

    /** Кодирует исходный [text] для перевода НА [targetLang]: `[__src__] tokens </s>`. */
    fun encodeSource(text: String, sourceLang: String): IntArray {
        val srcLangId = langTokenId(sourceLang)
        val body = encodeToIds(text)
        // TemplateProcessing single = [__src__] A </s>
        val ids = ArrayList<Int>(body.size + 2)
        ids += srcLangId
        ids += body
        ids += EOS_ID
        return ids.toIntArray()
    }

    /** Принудительный первый генерируемый токен при переводе на [targetLang] (например, `__ru__`). */
    fun targetLangId(targetLang: String): Int = langTokenId(targetLang)

    /** Id языкового кода точно по vocab, например `"ru"` → id токена `__ru__`. */
    fun langTokenId(lang: String): Int {
        val token = "__${lang.lowercase()}__"
        return vocab[token]
            ?: throw IllegalArgumentException("unknown M2M-100 language token: $token")
    }

    /** Декодирует сгенерированные id обратно в текст, выкидывая спец-/управляющие/языковые токены. */
    fun decode(ids: IntArray): String {
        val sb = StringBuilder()
        for (id in ids) {
            if (id == EOS_ID || id == PAD_ID || id == BOS_ID || id == unkId) continue
            val token = idToToken[id] ?: continue
            if (isLanguageToken(token)) continue
            sb.append(token)
        }
        // Декод metaspace: `▁` — граница слова → пробел; ведущую границу обрезаем.
        return sb.toString().replace(METASPACE, ' ').trim()
    }

    val eosId: Int get() = EOS_ID
    val padId: Int get() = PAD_ID
    val decoderStartId: Int get() = DECODER_START_ID

    // ---- внутренности ----------------------------------------------------------------------

    private fun encodeToIds(text: String): List<Int> {
        val normalized = Normalizer.normalize(text, Normalizer.Form.NFKC)
        val words = normalized.trim().split(WHITESPACE).filter { it.isNotEmpty() }
        val out = ArrayList<Int>()
        var lastWasUnk = false
        for (word in words) {
            val metaspaced = METASPACE + word
            for (piece in applyBpe(metaspaced)) {
                val id = vocab[piece]
                if (id != null) {
                    out += id
                    lastWasUnk = false
                } else if (!lastWasUnk) {
                    // Символа нет в vocab — fuse_unk сворачивает соседние неизвестные в один <unk>.
                    out += unkId
                    lastWasUnk = true
                }
            }
        }
        return out
    }

    /**
     * BPE для одного metaspace-слова. При `ignore_merges = true` слово, уже есть в vocab,
     * возвращается целиком; иначе режем на Unicode-кодпоинты (char coverage из SentencePiece) и
     * жадно сливаем соседнюю пару с минимальным рангом, пока такие пары есть.
     */
    private fun applyBpe(word: String): List<String> {
        if (vocab.containsKey(word)) return listOf(word)

        // Режем по Unicode-кодпоинтам (не по UTF-16 char), чтобы CJK/эмодзи не разваливались.
        val symbols = ArrayList<String>()
        var i = 0
        while (i < word.length) {
            val cp = word.codePointAt(i)
            val charCount = Character.charCount(cp)
            symbols += word.substring(i, i + charCount)
            i += charCount
        }
        if (symbols.size <= 1) return symbols

        while (true) {
            var bestRank = Int.MAX_VALUE
            var bestIndex = -1
            for (j in 0 until symbols.size - 1) {
                val rank = mergeRanks[pairKey(symbols[j], symbols[j + 1])] ?: continue
                if (rank < bestRank) {
                    bestRank = rank
                    bestIndex = j
                }
            }
            if (bestIndex < 0) break
            val merged = symbols[bestIndex] + symbols[bestIndex + 1]
            symbols[bestIndex] = merged
            symbols.removeAt(bestIndex + 1)
            if (symbols.size == 1) break
        }
        return symbols
    }

    private fun pairKey(left: String, right: String): Long {
        val l = vocab[left] ?: return ABSENT_PAIR
        val r = vocab[right] ?: return ABSENT_PAIR
        return (l.toLong() shl 32) or (r.toLong() and 0xFFFFFFFFL)
    }

    private fun isLanguageToken(token: String): Boolean =
        token.length >= 4 && token.startsWith("__") && token.endsWith("__")

    companion object {
        private const val TAG = "M2m100Tokenizer"
        const val METASPACE = '▁'
        private val WHITESPACE = Regex("\\s+")

        // Референсные управляющие id (стабильны для всего vocab M2M-100).
        private const val BOS_ID = 0
        private const val PAD_ID = 1
        private const val EOS_ID = 2
        // Декодер M2M-100 стартует с EOS id в роли decoder_start_token_id.
        private const val DECODER_START_ID = 2
        private const val ABSENT_PAIR = -1L

        /**
         * Загрузка из `tokenizer.json`. Возвращает null (никогда не бросает), если файла нет или он
         * структурно непригоден — чтобы вызывающий код мог честно заблокироваться.
         */
        fun load(tokenizerJson: File): M2m100Tokenizer? = runCatching {
            val root = JSONObject(tokenizerJson.readText())
            val model = root.getJSONObject("model")
            val vocabJson = model.getJSONObject("vocab")
            val vocab = HashMap<String, Int>(vocabJson.length() * 2)
            val idToToken = HashMap<Int, String>(vocabJson.length() * 2)
            val keys = vocabJson.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val id = vocabJson.getInt(key)
                vocab[key] = id
                idToToken[id] = key
            }
            // added_tokens (управляющие + языковые) переопределяют/дополняют базовый vocab.
            root.optJSONArray("added_tokens")?.let { added ->
                for (a in 0 until added.length()) {
                    val obj = added.getJSONObject(a)
                    val content = obj.getString("content")
                    val id = obj.getInt("id")
                    vocab[content] = id
                    idToToken[id] = content
                }
            }
            val merges = model.getJSONArray("merges")
            val mergeRanks = HashMap<Long, Int>(merges.length() * 2)
            for (m in 0 until merges.length()) {
                val entry = merges.get(m)
                val (left, right) = parseMerge(entry)
                val l = vocab[left]
                val r = vocab[right]
                if (l != null && r != null) {
                    mergeRanks[(l.toLong() shl 32) or (r.toLong() and 0xFFFFFFFFL)] = m
                }
            }
            val unkId = vocab["<unk>"] ?: 3
            Log.i(TAG, "tokenizer loaded: vocab=${vocab.size} merges=${mergeRanks.size}")
            M2m100Tokenizer(vocab, idToToken, mergeRanks, unkId)
        }.getOrElse { t ->
            Log.e(TAG, "tokenizer load failed", t)
            null
        }

        /** Записи merges — это либо строки "left right", либо массивы ["left","right"]. */
        private fun parseMerge(entry: Any): Pair<String, String> = when (entry) {
            is String -> {
                val sp = entry.indexOf(' ')
                entry.substring(0, sp) to entry.substring(sp + 1)
            }
            is org.json.JSONArray -> entry.getString(0) to entry.getString(1)
            else -> "" to ""
        }
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/M2m100Tokenizer.swift
import Foundation

// Faithful pure-Swift port of the M2M-100 HuggingFace fast tokenizer, driven entirely
// by the model's own `tokenizer.json`. Mirrors the Android M2m100Tokenizer exactly.
//
// Pipeline:
//  1. Normalize  — NFKC (Unicode). For RU/EN/ZH conversational text this is the faithful
//     subset of the reference SentencePiece Precompiled charsmap (width/compat folding).
//  2. Pre-tokenize — WhitespaceSplit then Metaspace: each word is prefixed with ▁ (U+2581)
//     and `add_prefix_space = true`.
//  3. BPE per word — `ignore_merges = true`: a word already in the vocab is emitted whole;
//     otherwise standard rank-ordered pair merging from codepoints up. `fuse_unk = true`
//     folds consecutive unknowns into a single <unk>. No dropout, no byte-fallback.
//  4. Post-process — TemplateProcessing `single`: `[__src__] tokens </s>`.
//
// Decoding: ▁ → space, trim leading/trailing whitespace, skip special tokens.
//
// Token ids come from the vocab map so language tokens (__en__=128022, __ru__=128077,
// __zh__=128102) and control ids (<s>=0, <pad>=1, </s>=2, <unk>=3) are the reference ids.
final class M2m100Tokenizer {
    // MARK: - Public constants
    static let metaspace: Character = "\u{2581}" // ▁

    static let bosId = 0
    static let padId = 1
    static let eosId = 2
    // M2M-100 decoder is seeded with EOS as decoder_start_token_id.
    static let decoderStartId = 2

    var eosId: Int { M2m100Tokenizer.eosId }
    var padId: Int { M2m100Tokenizer.padId }
    var decoderStartId: Int { M2m100Tokenizer.decoderStartId }

    // MARK: - Private state
    private let vocab: [String: Int]
    private let idToToken: [Int: String]
    // Key encodes a pair as (leftId << 32) | rightId; value is merge rank (lower = higher priority).
    private let mergeRanks: [UInt64: Int]
    private let unkId: Int

    private static let absentPair: UInt64 = UInt64.max

    // MARK: - Init
    private init(vocab: [String: Int], idToToken: [Int: String], mergeRanks: [UInt64: Int], unkId: Int) {
        self.vocab = vocab
        self.idToToken = idToToken
        self.mergeRanks = mergeRanks
        self.unkId = unkId
    }

    // MARK: - Public API

    // Encode source text for translation from sourceLang: [__src__] tokens </s>
    func encodeSource(text: String, sourceLang: String) -> [Int] {
        let srcLangId = langTokenId(lang: sourceLang)
        let body = encodeToIds(text: text)
        var ids = [Int]()
        ids.reserveCapacity(body.count + 2)
        ids.append(srcLangId)
        ids.append(contentsOf: body)
        ids.append(M2m100Tokenizer.eosId)
        return ids
    }

    // The forced first generated token for the target language, e.g. "ru" → id of __ru__.
    func targetLangId(lang: String) -> Int {
        langTokenId(lang: lang)
    }

    // Decode generated ids back to text: skip specials, convert ▁ to space, trim.
    func decode(ids: [Int]) -> String {
        var sb = ""
        for id in ids {
            if id == M2m100Tokenizer.eosId || id == M2m100Tokenizer.padId ||
               id == M2m100Tokenizer.bosId || id == unkId { continue }
            guard let token = idToToken[id] else { continue }
            if isLanguageToken(token) { continue }
            sb.append(token)
        }
        // ▁ marks word boundary → space; leading boundary trimmed.
        return sb.replacingOccurrences(of: "\u{2581}", with: " ").trimmingCharacters(in: .whitespaces)
    }

    // MARK: - Load

    // Load from a tokenizer.json. Returns nil (never throws) if the file is missing or unusable.
    static func load(from url: URL) -> M2m100Tokenizer? {
        guard let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let model = root["model"] as? [String: Any],
              let vocabJson = model["vocab"] as? [String: Any] else {
            return nil
        }

        var vocab = [String: Int](minimumCapacity: vocabJson.count * 2)
        var idToToken = [Int: String](minimumCapacity: vocabJson.count * 2)
        for (key, value) in vocabJson {
            guard let id = value as? Int else { continue }
            vocab[key] = id
            idToToken[id] = key
        }

        // added_tokens (control + language tokens) override/augment base vocab.
        if let addedTokens = root["added_tokens"] as? [[String: Any]] {
            for obj in addedTokens {
                guard let content = obj["content"] as? String,
                      let id = obj["id"] as? Int else { continue }
                vocab[content] = id
                idToToken[id] = content
            }
        }

        guard let mergesRaw = model["merges"] as? [Any] else { return nil }
        var mergeRanks = [UInt64: Int](minimumCapacity: mergesRaw.count * 2)
        for (rank, entry) in mergesRaw.enumerated() {
            guard let (left, right) = parseMerge(entry) else { continue }
            guard let l = vocab[left], let r = vocab[right] else { continue }
            let key = (UInt64(bitPattern: Int64(l)) << 32) | (UInt64(bitPattern: Int64(r)) & 0xFFFF_FFFF)
            mergeRanks[key] = rank
        }

        let unkId = vocab["<unk>"] ?? 3
        return M2m100Tokenizer(vocab: vocab, idToToken: idToToken, mergeRanks: mergeRanks, unkId: unkId)
    }

    // MARK: - Private helpers

    private func langTokenId(lang: String) -> Int {
        let token = "__\(lang.lowercased())__"
        guard let id = vocab[token] else {
            // Unknown language token — return unk. Callers gate on supported languages.
            return unkId
        }
        return id
    }

    private func encodeToIds(text: String) -> [Int] {
        // NFKC normalization mirrors Android's Normalizer.normalize(text, NFKC).
        let normalized = text.precomposedStringWithCompatibilityMapping
        let words = normalized.split(separator: " ", omittingEmptySubsequences: true)
            .map(String.init)
        var out = [Int]()
        var lastWasUnk = false
        let metaspaceStr = String(M2m100Tokenizer.metaspace)
        for word in words {
            let metaspaced = metaspaceStr + word
            for piece in applyBpe(word: metaspaced) {
                if let id = vocab[piece] {
                    out.append(id)
                    lastWasUnk = false
                } else if !lastWasUnk {
                    // fuse_unk: fold consecutive unknowns into one <unk>.
                    out.append(unkId)
                    lastWasUnk = true
                }
            }
        }
        return out
    }

    // BPE on a single metaspaced word. With `ignore_merges = true`, a word already in
    // the vocab is returned whole; otherwise split into Unicode scalars and merge greedily
    // by lowest rank until no more pairs are mergeable.
    private func applyBpe(word: String) -> [String] {
        if vocab[word] != nil { return [word] }

        // Split into Unicode scalar clusters (mirrors Android's codePointAt loop).
        var symbols = unicodeScalarClusters(word)
        if symbols.count <= 1 { return symbols }

        while true {
            var bestRank = Int.max
            var bestIndex = -1
            for j in 0 ..< symbols.count - 1 {
                if let rank = mergeRank(left: symbols[j], right: symbols[j + 1]) {
                    if rank < bestRank {
                        bestRank = rank
                        bestIndex = j
                    }
                }
            }
            if bestIndex < 0 { break }
            symbols[bestIndex] = symbols[bestIndex] + symbols[bestIndex + 1]
            symbols.remove(at: bestIndex + 1)
            if symbols.count == 1 { break }
        }
        return symbols
    }

    private func mergeRank(left: String, right: String) -> Int? {
        guard let l = vocab[left], let r = vocab[right] else { return nil }
        let key = (UInt64(bitPattern: Int64(l)) << 32) | (UInt64(bitPattern: Int64(r)) & 0xFFFF_FFFF)
        return mergeRanks[key]
    }

    // Split a string into individual Unicode scalar value strings (preserves surrogate pairs / CJK).
    private func unicodeScalarClusters(_ s: String) -> [String] {
        s.unicodeScalars.map { String($0) }
    }

    private func isLanguageToken(_ token: String) -> Bool {
        token.count >= 4 && token.hasPrefix("__") && token.hasSuffix("__")
    }

    // Parse a merge entry: either "left right" String or ["left","right"] Array.
    private static func parseMerge(_ entry: Any) -> (String, String)? {
        if let s = entry as? String {
            guard let spaceIdx = s.firstIndex(of: " ") else { return nil }
            let left = String(s[s.startIndex ..< spaceIdx])
            let right = String(s[s.index(after: spaceIdx)...])
            return (left, right)
        }
        if let arr = entry as? [String], arr.count >= 2 {
            return (arr[0], arr[1])
        }
        return nil
    }
}
